Adopt stereo content component for the model's portal entity

achan00 opened the pull request:
#### 41a4bbb26f95103dfe66e08600b2606ce69b17c6
<pre>
Adopt stereo content component for the model&apos;s portal entity
<a href="https://bugs.webkit.org/show_bug.cgi?id=288019">https://bugs.webkit.org/show_bug.cgi?id=288019</a>
<a href="https://rdar.apple.com/120925261">rdar://120925261</a>

Reviewed by Mike Wyrzykowski.

With this adoption, the portal layer is not separated
and no longer on top of all other content.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/platform/graphics/Color.h:
* Source/WebKit/UIProcess/ios/WKPageHostedModelView.mm:
(+[WKPageHostedModelView _usesStereoContent]):
Check runtime flag to determine whether to use stereo
content component. Default values are different between
device and simulator.
(-[WKPageHostedModelView init]):
If using stereo content, set up a new CALayer to associate
with the stereo content entity. A world root is set up
with the stereo content component. The stereo content
entity is now set as the parent of the container view&apos;s entity.
(-[WKPageHostedModelView dealloc]):
Make sure the stereo content entity is removed from the scene.
(-[WKPageHostedModelView setShouldDisablePortal:]):
To disable the portal, remove the stereo content layer and entity
from the hierarchy. To show the portal, add the stereo content
layer and entity back.
(-[WKPageHostedModelView applyBackgroundColor:]):
Update the background color of the portal via
REEmbeddedStereoContentComponentSetClearColor(). Since
that function expects linear RGB, we need to convert
the sRGB color to linear first.
(-[WKPageHostedModelView layoutSubviews]):
Make sure the stereo content layer&apos;s frame gets updated
with its parent layer.
* Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
(WKSRKEntity.applyIBL(_:completion:)):
This version check should be ok to remove now.

Canonical link: <a href="https://commits.webkit.org/291371@main">https://commits.webkit.org/291371@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/602760db9fe0753edc4b98e6900711753fa7a06e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/91741 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/11338 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/814 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/96696 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/42386 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/93791 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/11640 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/19896 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/70434 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/27931 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/94742 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/8942 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/83116 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/50760 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/8689 "Found 1 new test failure: imported/w3c/web-platform-tests/svg/struct/scripted/blank.svg (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/708 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/41582 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/84537 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/78958 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/709 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/98717 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/90483 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/18885 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/14041 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/79465 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/19138 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/78960 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/78690 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19667 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/23200 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/536 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/11964 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/18875 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/24116 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/113071 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/18578 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/32748 "Found 12 new JSC stress test failures: microbenchmarks/memcpy-wasm-medium.js.default, microbenchmarks/memcpy-wasm-medium.js.dfg-eager, microbenchmarks/memcpy-wasm-small.js.bytecode-cache, microbenchmarks/memcpy-wasm-small.js.dfg-eager, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.default-wasm, wasm.yaml/wasm/function-tests/memcpy-wasm-loop.js.wasm-eager, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.wasm-collect-continuously, wasm.yaml/wasm/stress/b3-signed-extend-16-to-64.js.wasm-eager-jettison, wasm.yaml/wasm/stress/ipint-bbq-osr-with-try5.js.wasm-eager, wasm.yaml/wasm/stress/live-funcref-across-loop-tier-up.js.wasm-eager ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/22032 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/20332 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/ios/WKPageHostedModelView.mm:
```diff
@@ -36,15 +36,49 @@
+#define HAVE_RE_STEREO_CONTENT_SUPPORT RE_FEATURE_EMBEDDED_STEREO_CONTENT_COMPONENT
+#else
+#define HAVE_RE_STEREO_CONTENT_SUPPORT 0
+#endif
```
This should be defined in `wtf/PlatformHave.h` and then used in this file as  `HAVE(RE_STEREO_CONTENT_SUPPORT)`.

mwyrzykowski commented on Source/WebKit/UIProcess/ios/WKPageHostedModelView.mm:
```diff
@@ -36,15 +36,46 @@
 #import <pal/spi/cocoa/QuartzCoreSPI.h>
 #import <wtf/RetainPtr.h>
 
+#if __has_include(<CoreRE/CoreRE_SPI_WebKit.h>)
```
I suppose we could use `HAVE(RE_STEREO_CONTENT_SUPPORT)`? But does not really matter to me

achan00 replied:
Maybe - but I figure it's more straightforward and reliable to just check for the header existence before including it. Do let me know if there's a good reason to do it a different way.

mwyrzykowski commented on Source/WebKit/UIProcess/ios/WKPageHostedModelView.mm:
```diff
@@ -149,6 +252,32 @@ - (void)setShouldDisablePortal:(BOOL)shouldDisablePortal
+        if (backgroundColor && backgroundColor->isValid()) {
+            auto opaqueColor = backgroundColor->opaqueColor();
+            auto [r, g, b, a] = opaqueColor.toResolvedColorComponentsInColorSpace(WebCore::ColorSpace::LinearSRGB);
+            clearColorComponents = simd_make_float3(r, g, b);
```
Do we want to multiply the alpha into this clear color or does it not matter? I.e., should it be:
```
clearColorComponents = simd_make_float3(r * a, g * a, b * a);
```
or maybe not?

achan00 replied:
I talked with Brandel and I think it's ok to leave it as is, especially since the color passed in is already opaque.

achan00 committed:
Adopt stereo content component for the model's portal entity
https://bugs.webkit.org/show_bug.cgi?id=288019
rdar://120925261

Reviewed by Mike Wyrzykowski.

With this adoption, the portal layer is not separated
and no longer on top of all other content.

* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/platform/graphics/Color.h:
* Source/WebKit/UIProcess/ios/WKPageHostedModelView.mm:
(+[WKPageHostedModelView _usesStereoContent]):
Check runtime flag to determine whether to use stereo
content component. Default values are different between
device and simulator.
(-[WKPageHostedModelView init]):
If using stereo content, set up a new CALayer to associate
with the stereo content entity. A world root is set up
with the stereo content component. The stereo content
entity is now set as the parent of the container view's entity.
(-[WKPageHostedModelView dealloc]):
Make sure the stereo content entity is removed from the scene.
(-[WKPageHostedModelView setShouldDisablePortal:]):
To disable the portal, remove the stereo content layer and entity
from the hierarchy. To show the portal, add the stereo content
layer and entity back.
(-[WKPageHostedModelView applyBackgroundColor:]):
Update the background color of the portal via
REEmbeddedStereoContentComponentSetClearColor(). Since
that function expects linear RGB, we need to convert
the sRGB color to linear first.
(-[WKPageHostedModelView layoutSubviews]):
Make sure the stereo content layer's frame gets updated
with its parent layer.
* Source/WebKit/WebKitSwift/RealityKit/RKEntity.swift:
(WKSRKEntity.applyIBL(_:completion:)):
This version check should be ok to remove now.

Canonical link: https://commits.webkit.org/291371@main
